src/app/qgisapp.cpp

```diff
@@ -7201,6 +7201,45 @@ void QgisApp::saveAsFile( QgsMapLayer *layer, bool onlySelected )
   }
 }
 
+void QgisApp::makeMemoryLayerPermanent( QgsVectorLayer *layer )
+{
+  if ( !layer )
+    return;
+
+  const QString layerId = layer->id();
+
+  auto onSuccess = [this, layerId]( const QString & newFilename,
+                                    bool,
+                                    const QString &,
+                                    const QString &,
+                                    const QString & )
+  {
+    // we have to re-retrieve the layer, in case it's been removed during the lifetime of the writer task
+    QgsVectorLayer *vl = qobject_cast< QgsVectorLayer * >( QgsProject::instance()->mapLayer( layerId ) );
+    if ( vl )
+    {
+      QgsDataProvider::ProviderOptions options;
+      vl->setDataSource( QStringLiteral( "%1" ).arg( newFilename ), vl->name(), QStringLiteral( "ogr" ), options );
+      this->messageBar()->pushMessage( tr( "Layer Saved" ),
+                                       tr( "Successfully saved scratch layer to <a href=\"%1\">%2</a>" ).arg( QUrl::fromLocalFile( QFileInfo( newFilename ).path() ).toString(), QDir::toNativeSeparators( newFilename ) ),
```

Comment:
Don't use the path as hyperlink, use the file, it'll benefit from our recent native file opening work :)